### PR TITLE
feat: #49 /login endpoint

### LINKED ISSUE #49: Support Wallet Login & Lock/Unlock
## hub-auth resources

<details><summary>View</summary>

resource|type|sensitive?|served?|usage
--------|----|----------|-------|-----
key server URLs|list of URLs|no|yes|edge-agent Vue.js sets up key stores, then initializes wallet
sds server URLs|list of URLs|no|yes|edge-agent Vue.js sets up vaults, then initializes wallet
keyStoreIDs|list of IDs|yes|yes|edge-agent aries-framework-go KMS
SDS vault IDs|list of IDs|yes|yes|edge-agent aries-framework-go storage
user devices info|list of tuples {AAGUID publicKey, x509_cert}|yes|no|when unlocking: validate if device is registered and to which user before giving access to bootstrap resources (primary key is wiped from devices when locked, needs to be fetched from network)
locations of user primary key shares|list of SDS locations of the user's primary key shares (split w/SSS)|yes|yes|a) when unlocking: edge-agent Vue.js needs location + authZ to fetch those shares; b) key server needs locations + authZ to fetch those shares to unlock the user's key store and do crypto operations
authorized wallet callback URIs|list of URIs|no|no|configuration: helps restrict processing of authZ requests to authorized wallets
</details>

## Breakdown

<details><summary>View</summary>

- [ ] hub-auth
  - [ ] /login endpoint #50 
    - [ ] serve page with OIDC provider options
    - [ ] invokes `/oauth2/request` endpoint and redirects
   - [ ] `handleOIDCCallback()` should accept login request at hydra and redirect back to hydra
  - [ ] /consent endpoint
    - [ ] accept consent request at hydra and redirect back to hydra
  - [ ] BDD Tests with Hydra #52
  - [ ] setup Hydra
  - [ ] BDD Tests
  - [ ] GET /bootstrap
    - [ ] protect with access_token
    - [ ] introspect at Hydra - use `sub` to locate the user's bootstrap data
  - [ ] PUT /bootstrap
    - [ ] protect with access_token
    - [ ] updates user's bootstrap data
  - [ ] configure allowed user devices
    - [ ] TODO collection of tuples: {AAGUID, x509_cert} ?
  - [ ] POST /device
    - [ ] protect with access_token
    - [ ] TODO model: {AAGUID, x509_cert} ?
    - [ ] validate: AAGUID/x509_cert are allowed
- [ ] session expiry
  - [ ] expiry of access_tokens #55 
</details>

### LINKED ISSUE #50: hydra login handler
part of #49 

Add a `/login` endpoint to handle Hydra's login request challenge.

### PR DESCRIPTION
**Pull request by llorllale**
closes #50 

* Added `/hydra/login`
* Added hydra url startup param
* Pulled in latest fixes from edge-core
* Moved some parts of the code around a bit

Signed-off-by: George Aristy <george.aristy@securekey.com>

### CODE REVIEW

**Comment by llorllale** [comment]
Can we disregard the red flag on coverage?

* of all new lines of code, only 1 is not tested because the code is unreachable (transient.go#L27)
* all other untested lines are existing code that I moved around

**Review by talwinder50 — APPROVED** [review]

File: pkg/restapi/operation/operations.go [hunk 1]
```diff
@@ -204,19 +142,57 @@ func New(config *Config) (*Operation, error) {
...
 
-	return provider.OpenStore(bootstrapStoreName)
+	req := admin.NewGetLoginRequestParams()
+
+	req.SetLoginChallenge(challenge)
+
+	login, err := c.hydra.GetLoginRequest(req)
+	if err != nil {
+		common.WriteErrorResponsef(w, logger,
+			http.StatusBadGateway, "failed to fetch login request from hydra: %s", err.Error())
```
Comment (DRK3) [inline pkg/restapi/operation/operations.go:172]: Is http.StatusBadGateway the appropriate header for all errors returned from GetLoginRequest?

File: pkg/restapi/operation/operations_test.go [hunk 2]
```diff
@@ -17,6 +17,9 @@ import (
 	"strings"
 	"testing"
 
+	"github.com/ory/hydra-client-go/client/admin"
+	"github.com/ory/hydra-client-go/models"
+
```
Comment (DRK3) [inline pkg/restapi/operation/operations_test.go:22]: Extra line break

File: test/bdd/fixtures/auth-rest/docker-compose.yml [hunk 3]
```diff
@@ -23,6 +23,7 @@ services:
       - AUTH_REST_GOOGLE_CLIENTSECRET=TODO
       - AUTH_REST_SDS_URL=TODO        # onboard user: https://github.com/trustbloc/hub-auth/issues/38
       - AUTH_REST_KEYSERVER_URL=TODO  # onboard user: https://github.com/trustbloc/hub-auth/issues/38
+      - AUTH_REST_HYDRA_URL=http://TODO.com      # setup Hydra in BDD tests: https://github.com/trustbloc/hub-auth/issues/52
```
Comment (DRK3) [inline test/bdd/fixtures/auth-rest/docker-compose.yml:26]: I would add the TODO word here to be consistent with the other TODOs
(and I guess that would apply to the TODOs above these two lines as well...)

**Review by DRK3 — APPROVED** [review]

**Comment by DRK3** [comment]
@llorllale Only a few minor things to take a quick look at - nothing major. Can be merged

File: pkg/restapi/operation/operations.go (hunk 1 above)
Comment (llorllale) [reply pkg/restapi/operation/operations.go:172]: Good question. I don't know. There is a `login.Error() string` method... I guess we can figure that out later.

File: test/bdd/fixtures/auth-rest/docker-compose.yml (hunk 3 above)
Comment (llorllale) [reply test/bdd/fixtures/auth-rest/docker-compose.yml]: fixed

**Pull request merged**